WordPress/gutenberg
# Block Hooks: Set ignoredHookedBlocks metada attr upon insertion

## What?
Set the `metadata.ignoredHookedBlocks` attribute upon inserting a new block that has hooked blocks.

## Why?
To unblock https://github.com/WordPress/wordpress-develop/pull/5726 (see https://github.com/WordPress/wordpress-develop/pull/5712#issuecomment-1838758546).

In WP 6.5, we'll be enabling insertion of hooked blocks into modified templates/parts/patterns. This is implemented by adding a `metadata.ignoredHookedBlocks` attribute to the hooked block's anchor block upon editing.

As a consequence, upon insertion of an anchor block, this attribute also needs to be added to that newly inserted instance.

_A note on UX: It could be argued that ideally, insertion of a new anchor block should cause its hooked blocks to be inserted alongside with it. This could be a future addition but seems to be a bit more complicated to do. Regardless, the `metadata.ignoredHookedBlocks` attribute also would need to be set in that case, in order to avoid double insertion -- meaning that adding that attribute is required as a baseline anyway._

## How?

AFAICT, there are two mechanisms that set attributes for a newly inserted block:
- `buildBlockTypeItem` is used to build items that appear in the inserter; conveniently, it allows setting an `initialAttribute`.
- OTOH, blocks that _don't_ show up in the inserter but are inserted as inner blocks of another block (e.g. the Comment Template block, as a newly inserted Comments block's inner block). These blocks are typically sync'd via `synchronizeBlocksWithTemplate` from a `template` set in the container block.

Both mechanisms are modified by this PR to set the `metadata.ignoredHookedBlocks` attribute.

## Testing Instructions
- Install and activate the [latest version of the Like Button plugin](https://github.com/ockham/like-button/releases/latest).
- Use a block theme like TT3 or TT4.
- In the Site Editor, edit the Single Posts template.
- Verify that the Like Button block is inserted both below the Post Content block, and as the Comment Template block's last child.
- Insert a new instance of the Post Content block right above the footer template part.
- Switch to the Code Editor and verify that the newly inserted instance of the Post Content block has its `metadata.ignoredHookedBlocks` attribute set to include the Like Button block: ``.
- Delete that block, and insert a new instance of the Comments block.
- Switch to the Code Editor and verified that the Comment Template block inside the new Comments block has its `metadata.ignoredHookedBlocks` attribute set to include the Like Button block: ``

## Follow-up

Currently, only hooked blocks that were added via `block.json` will be taken into account, but not the ones added via a filter. This will be addressed by #58622.

We should then also use it the selector to compute `hookedBlocksForCurrentBlock` in `block-hooks.js` (so it'll also use the new endpoint as its source of truth). (I've tried that out in https://github.com/WordPress/gutenberg/pull/58553/commits/6f973f9859b8522306268c6a5d6740bd8eec66b8 but decided to punt for now as it'd add needless complexity for the time being.)
